util/settings/settings.go

```diff
@@ -689,19 +689,28 @@ func (mgr *SettingsManager) SaveSSHKnownHostsData(ctx context.Context, knownHost
 		return err
 	}
 
-	// If we are running outside a K8S cluster, the ConfigMap mount will not
-	// automatically write out the SSH known hosts data. We need this mechanism
-	// for running E2E tests.
+	return mgr.ResyncInformers()
+}
+
+// If we are running outside a K8S cluster, the ConfigMap mount will not
+// automatically write out the SSH known hosts data. We need this mechanism
+// for running E2E tests.
+func (mgr *SettingsManager) SyncSSHKnownHostsData(ctx context.Context) {
 	if os.Getenv(common.EnvVarFakeInClusterConfig) == "true" {
+		certCM, err := mgr.GetConfigMapByName(common.ArgoCDKnownHostsConfigMapName)
+		if err != nil {
+			log.Errorf("Could not get ConfigMap %s: %v", common.ArgoCDKnownHostsConfigMapName, err)
+			return
+		}
+		knownHosts := certCM.Data["ssh_known_hosts"]
 		knownHostsPath := certutil.GetSSHKnownHostsDataPath()
-		err := ioutil.WriteFile(knownHostsPath, []byte(sshKnownHostsData), 0644)
+		log.Debugf("Writing known hosts data to %s", knownHostsPath)
```

Comment:
I would prefer to avoid adding test-related logic into production code. Instead can you implement code in test infrastructure which watches config map and sync in to file?

For example we can add following to Procfile:

```
for hosts in `kubectl get configmap argocd-ssh-known-hosts-cm -w  -o=jsonpath={.data.ssh_known_hosts}`; do echo $hosts > /tmp/knownhosts; done
```

Reply:
Please see comment @ https://github.com/argoproj/argo-cd/pull/2101#issuecomment-518725946